src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -429,72 +366,64 @@ private function createNewInvoice(Order $order, Cart $cart, $isFreeShipping, arr
         );
         $carrier = new Carrier((int) $order->id_carrier);
         $taxCalculator = $carrier->getTaxCalculator($invoice_address);
-
-        $invoice->total_paid_tax_excl = Tools::ps_round(
-            (float) $cart->getOrderTotal(false, $totalMethod, $newProducts),
-            $this->computingPrecision
-        );
-        $invoice->total_paid_tax_incl = Tools::ps_round(
-            (float) $cart->getOrderTotal(true, $totalMethod, $newProducts),
-            $this->computingPrecision
-        );
-        $invoice->total_products = (float) $cart->getOrderTotal(false, Cart::ONLY_PRODUCTS, $newProducts);
-        $invoice->total_products_wt = (float) $cart->getOrderTotal(true, Cart::ONLY_PRODUCTS, $newProducts);
-        $invoice->total_shipping_tax_excl = (float) $cart->getTotalShippingCost(null, false);
-        $invoice->total_shipping_tax_incl = (float) $cart->getTotalShippingCost();
-
-        $invoice->total_wrapping_tax_excl = abs($cart->getOrderTotal(false, Cart::ONLY_WRAPPING, $newProducts));
-        $invoice->total_wrapping_tax_incl = abs($cart->getOrderTotal(true, Cart::ONLY_WRAPPING, $newProducts));
         $invoice->shipping_tax_computation_method = (int) $taxCalculator->computation_method;
+        // We don't need to compute the invoice totals because the OrderAmountUpdater will take care of it afterwards
         $invoice->add();
 
         $invoice->saveCarrierTaxCalculator($taxCalculator->getTaxesAmount($invoice->total_shipping_tax_excl));
 
-        $orderCarrier = new OrderCarrier();
-        $orderCarrier->id_order = (int) $order->id;
-        $orderCarrier->id_carrier = (int) $order->id_carrier;
-        $orderCarrier->id_order_invoice = (int) $invoice->id;
-        $orderCarrier->weight = (float) $cart->getTotalWeight();
-        $orderCarrier->shipping_cost_tax_excl = (float) $invoice->total_shipping_tax_excl;
-        $orderCarrier->shipping_cost_tax_incl = (float) $invoice->total_shipping_tax_incl;
-        $orderCarrier->add();
+        if (true === $withFreeShipping) {
+            $this->addFreeShippingDiscount($order, $cart, $invoice);
+        }
 
         return $invoice;
     }
 
     /**
-     * @param int $orderInvoiceId
+     * @param Order $order
      * @param Cart $cart
-     * @param array $newProducts
+     * @param OrderInvoice $orderInvoice
      *
-     * @return OrderInvoice
+     * @throws OrderException
      */
-    private function updateExistingInvoice($orderInvoiceId, Cart $cart, array $newProducts)
+    private function addFreeShippingDiscount(Order $order, Cart $cart, OrderInvoice $orderInvoice): void
     {
-        $invoice = new OrderInvoice($orderInvoiceId);
+        $defaultLangId = (int) Configuration::get('PS_LANG_DEFAULT');
+        $cartRuleObj = new CartRule();
```

Comment:
Maybe we could use a factory here, so we can reuse this logic. I have already seen it multiple times in PS codebase.
```suggestion
        $cartRuleObj = CartRule::createFreeShippingCartRule();
```

Reply:
I had similar questions indeed, but not specifically on the Free shipping cart rule because I'm not sure there are that many places where we specifically need this (usually we create a cart rule based on a switch type: percent, amount, free shipping)

However I though of having a protected function to easily create a CartRule that is specifically related to an order/invoice, as it needs to have some specific settings (related to customer, limited number of uses, not visible, ...) and could be common with the `AddCartRuleToOrderHandler`